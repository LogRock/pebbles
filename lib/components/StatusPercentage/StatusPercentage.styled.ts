import styled from "styled-components";
import { StatusPercentageProps } from "./StatusPercentage.types";

export const StatusPctWrapper = styled.div<Pick<StatusPercentageProps, "size">>`
  display: flex;
  align-items: center;
  justify-content: flex-start;

  font-family: ${({ theme }) => theme.statusPercentage.fontFamily};
  font-size: ${({ theme, size }) => theme.statusPercentage.fontSize[size]};
  font-weight: ${({ theme }) => theme.statusPercentage.fontWeight};
  text-align: center;
`;
