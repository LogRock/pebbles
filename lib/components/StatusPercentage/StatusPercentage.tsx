import React, { FC } from "react";

import ProgressRing from "../Loader/ProgressRing";
import { StatusPctWrapper } from "./StatusPercentage.styled";
import { StatusPercentageProps } from "./StatusPercentage.types";

export interface StatusPercentageSizeType {
  radius: number;
  stroke: number;
}

const StatusPercentage: FC<StatusPercentageProps> = ({
  percentage,
  size = "medium",
  children,
  ...rest
}) => {
  const sizes = {
    small: { radius: 16, stroke: 4 },
    medium: { radius: 20, stroke: 4 },
    large: { radius: 24, stroke: 5 },
  };

  return (
    <StatusPctWrapper size={size} {...rest}>
      <ProgressRing
        percentage={percentage}
        radius={sizes[size].radius}
        stroke={sizes[size].stroke}
        showBaseRing
      />
      {children}
    </StatusPctWrapper>
  );
};

export default StatusPercentage;
