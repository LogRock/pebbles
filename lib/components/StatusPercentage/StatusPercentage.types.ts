import { spacingTokens } from "../../types/tokens";

export interface StatusPercentageProps {
  percentage: number;
  size: "small" | "medium" | "large";
  variant:
    | "primary"
    | "secondary"
    | "success"
    | "destructive"
    | "warning"
    | "neutral";
  spaceAfter?: spacingTokens;
}
