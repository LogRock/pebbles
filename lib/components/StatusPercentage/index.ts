export { default } from "./StatusPercentage";
export type { StatusPercentageProps } from "./StatusPercentage.types";
