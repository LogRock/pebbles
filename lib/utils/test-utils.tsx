/* eslint-disable import/export */
import React from "react";
import { cleanup, render } from "@testing-library/react";
import { afterEach } from "vitest";
import { ThemeProvider } from "styled-components";
import { light as theme } from "../theme";

afterEach(() => {
  cleanup();
});

const customRender = (ui: React.ReactElement, options = {}) =>
  render(ui, {
    // wrap provider(s) here if needed
    wrapper: ({ children }) => (
      <ThemeProvider theme={theme}>{children}</ThemeProvider>
    ),
    ...options,
  });

export * from "@testing-library/react";
export { default as userEvent } from "@testing-library/user-event";
// override render export
export { customRender as render };
