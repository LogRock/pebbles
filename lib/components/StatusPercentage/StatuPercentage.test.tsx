import React from "react";
import { render, screen } from "../../utils/test-utils";
import StatusPercentage from "./StatusPercentage";

describe("StatusPercentage tests", () => {
  it("shows main UI element", () => {
    const { container } = render(
      <StatusPercentage variant="success" percentage={50} size="medium">
        50
      </StatusPercentage>
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  it("shows correct percentage number on the component", () => {
    render(
      <StatusPercentage variant="success" percentage={50} size="medium">
        50
      </StatusPercentage>
    );
    screen.debug();
    expect(screen.getByText("50")).toBeInTheDocument();
  });

  it("renders progress ring with correct green color", () => {
    render(
      <StatusPercentage variant="success" percentage={50} size="medium">
        50
      </StatusPercentage>
    );
    screen.debug();
    expect(screen.getByTestId("ring")).toHaveAttribute("stroke", "#D4F9E1");
  });
});
