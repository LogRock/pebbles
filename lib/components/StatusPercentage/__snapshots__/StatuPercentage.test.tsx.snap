// Vitest Snapshot v1

exports[`StatusPercentage tests > shows main UI element 1`] = `
<div
  class="sc-dkPtRN hlCbZg"
>
  <div
    class="sc-bdvvtL iLkRqh"
    size="40"
  >
    <svg
      height="40"
      width="40"
    >
      <circle
        cx="20"
        cy="20"
        data-testid="ring"
        fill="transparent"
        r="12"
        stroke="#D4F9E1"
        stroke-width="4"
        style="stroke-dashoffset: 37.69911184307752;"
      />
    </svg>
    <svg
      height="40"
      width="40"
    >
      <defs>
        <lineargradient
          gradientUnits="userSpaceOnUse"
          id="ring_gradient"
          x1="0"
          x2="71.8942"
          y1="0"
          y2="10.6352"
        >
          <stop
            stop-color="#FF5F6D"
          />
          <stop
            offset="1"
            stop-color="#FFC371"
          />
        </lineargradient>
      </defs>
      <circle
        cx="20"
        cy="20"
        fill="transparent"
        r="12"
        stroke="#1BB954"
        stroke-dasharray="75.39822368615503 75.39822368615503"
        stroke-width="4"
        style="stroke-dashoffset: 37.69911184307752;"
      />
    </svg>
  </div>
  50
</div>
`;
